### ISSUE TITLE
[BUG] Double file scan with stats skipping

### ISSUE THREAD

**Issue by Kimahriman** [delta-io/delta#1073]
### Describe the problem
File stats skipping is causing two `filesForScan` operations.

#### Steps to reproduce
Haven't been able to make a reproduction, trying to. Only have hit this in a production table.

#### Observed results
There are two scan stages, caused by hitting [this](https://github.com/delta-io/delta/blob/f598157fb188daa32b1c5912ce2599feab24984e/core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala#L326) line of code. However, an example output is:

```
Prepared scan does not match actual filters. Reselecting files to query.
Prepared: ExpressionSet((raw#437.server_timestamp = 2020-12-08 05:59:20))
Actual: ExpressionSet((raw#437.server_timestamp = 2020-12-08 05:59:20))
```

#### Expected results
The expression sets are the same so it shouldn't trigger a second file scan.

### Environment information

* Delta Lake version: 1.2.0
* Spark version: 3.2.1
* Scala version: 2.12

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [ ] No. I cannot contribute a bug fix at this time.

**Comment by allisonport-db** [comment]
Hi @Kimahriman thanks for reporting this! This is definitely weird and I wasn't able to reproduce it. Could you update this issue with more details if you are able to reproduce it?

**Comment by Kimahriman** [comment]
Yeah I'll keeping digging into and see if I can find what's triggering this

**Comment by Kimahriman** [comment]
Ok figured some things out. There seems to be two issues when using nested columns.
1. `prepared` filters are pre-schema pruning, and `actual` filters are after. This can make the `ordinal` value of `GetStructField` different when comparing expressions
2. The `actual` seems to include the optional `name` of `GetStructField`, while `prepared` doesn't

Sample code to reproduce:
```
from pyspark.sql import SparkSession
import pyspark.sql.functions as F

spark = SparkSession.builder.getOrCreate()

df = spark.range(5).select(F.struct(F.lit('test').alias('test'), F.col('id').cast("string").alias('id')).alias('nested'))
df.write.format('delta').save('test')

spark.read.format('delta').load('test').filter('nested.id = "2"').count()
```

And with some extra logging where I output the contents of `GetStructField`:

```
22/04/17 08:52:09 INFO PreparedDeltaFileIndex: 
Prepared scan does not match actual filters. Reselecting files to query.
Prepared: ExpressionSet((nested#457.id = 2), isnotnull(nested#457))
Actual: ExpressionSet(isnotnull(nested#457), (nested#457.id = 2))
         
22/04/17 08:52:09 INFO PreparedDeltaFileIndex: Prepared:
22/04/17 08:52:09 INFO PreparedDeltaFileIndex: nested#457, 1, Some(id)
22/04/17 08:52:09 INFO PreparedDeltaFileIndex: Actual:
22/04/17 08:52:09 INFO PreparedDeltaFileIndex: nested#457, 0, None
```

**Comment by Kimahriman** [comment]
What's the case where the actual is supposed to possibly be different than prepared?

**Comment by Kimahriman** [comment]
Likewise, what's the use case of pre-calculating the list of files? It seems weird to get the list of files as part of the optimization process, before the physical plan is even built.

**Comment by tdas** [comment]
Pre calculating the list of files before the "optimization" step actually helps reduce the file list and generate better stats for the optimizer to optimize plans (join reordering, etc.).

However, whats happening here is that the optimizer is changing the predicates which is invalidating the calculated list. It seems in some cases the list is still valid as the predicates seem to be conceptually same, so the  list should be usable. Its tricky to find out whether the new predicates are same as before or not, hence we are always blindly falling back. There is clearly a room for improvement, but we have to do very cautiously as we do not want to compute incorrect results.

**Comment by tdas** [comment]
Follow up qs, do you see this problem with filters on nested fields or on top-level fields as well. trying to understand the scope of the problem.

**Comment by Kimahriman** [comment]
It should be purely on nested fields (we basically only have nested fields), because the schema won't change for a non-nested field (just other top level fields will be pruned). However I see this happening on almost any ad-hoc query of nested data, since you are probably always pruning something.

**Comment by tdas** [comment]
I got confuse with that. Its only the fields in  the filter that should matter here. And you are saying that filter on any nesteed field always causes double scans.. right? And if so, have you tried a filter on a top-level field? Did it avoid double scan?

**Comment by Kimahriman** [comment]
Maybe just transform `GetStructField` to some `UnresolvedAttribute`-like expression before comparing the two sets?

**Comment by foss-contributor** [comment]
@Kimahriman, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!

**Comment by felipepessoto** [comment]
PR #1096 is still open
